pilot/adapter/config/file/monitor.go

```diff
@@ -0,0 +1,390 @@
+	}
+}
+
+func (m *Monitor) readFiles() resources {
+	var result resources
+
+	err := filepath.Walk(m.root, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if !supportedExtensions[filepath.Ext(path)] || (info.Mode()&os.ModeType) != 0 {
+			return nil
+		}
+		data, err := ioutil.ReadFile(path)
+		if err != nil {
+			glog.Warningf("Failed to read %m: %v", path, err)
+			return err
+		}
+		for _, r := range parseYaml(data) {
+			if !m.kinds[r.Kind] {
+				// Ignore unsupported kinds
+				continue
+			}
+			result = append(result, r)
+		}
+		return nil
+	})
+	if err != nil {
+		glog.Errorf("failure during filepath.Walk: %v", err)
+	}
+
+	// Sort the resources by their IDs.
+	result.Sort()
+
+	return result
+}
+
+// ParseYamlFile is a utility method that parses the given yaml file and returns all of the Istio configs. This is
+// equivalent to calling ioutil.ReadPath, followed by ParseYaml.
+func ParseYamlFile(path string) ([]*model.Config, error) {
+	data, err := ioutil.ReadFile(path)
+	if err != nil {
+		return nil, err
+	}
+
+	return ParseYaml(data)
+}
+
+// ParseYaml is a utility method that parses the contents of a given yaml file and returns all of the Istio configs.
+func ParseYaml(data []byte) ([]*model.Config, error) {
+	resources := parseYaml(data)
+	if len(resources) == 0 {
+		return nil, nil
+	}
+
+	configs := make([]*model.Config, len(resources))
+	for i, r := range resources {
+		var err error
+		if configs[i], err = r.toConfig(); err != nil {
+			return nil, err
+		}
+	}
+
+	return configs, nil
+}
+
+// parseYaml parses the data for an entire yaml file and returns slices for individual resources.
+func parseYaml(data []byte) []*resource {
+	if bytes.HasPrefix(data, []byte("---\n")) {
+		data = data[4:]
+	}
+	if bytes.HasSuffix(data, []byte("\n")) {
+		data = data[:len(data)-1]
+	}
+	if bytes.HasSuffix(data, []byte("\n---")) {
+		data = data[:len(data)-4]
+	}
+	if len(data) == 0 {
+		return nil
+	}
+	chunks := bytes.Split(data, []byte("\n---\n"))
+	resources := make([]*resource, 0, len(chunks))
+	for i, chunk := range chunks {
+		r, err := parseResource(chunk)
+		if err != nil {
+			glog.Errorf("Error processing config [%d]: %v", i, err)
+			continue
+		}
+		if r == nil {
+			continue
+		}
+		resources = append(resources, r)
+	}
+	return resources
+}
+
+// parseResource parses a subsection of yaml for a single resource.
+func parseResource(data []byte) (*resource, error) {
+	r := &resource{}
+	if err := yaml.Unmarshal(data, r); err != nil {
+		return nil, err
+	}
+	if empty(r) {
+		// Can be empty if no data or if it just contains comments.
+		return nil, nil
+	}
+	if r.Kind == "" && r.Metadata.Namespace == "" && r.Metadata.Name == "" {
+		return nil, fmt.Errorf("key elements are empty.\n <<%s>>", string(data))
+	}
+	r.sha = sha1.Sum(data)
+	return r, nil
+}
+
+// Check if the parsed resource is empty
+func empty(r *resource) bool {
+	return reflect.DeepEqual(*r, *emptyResource)
```

Comment:
Just use `&resource{}` inline here, no need for the (package) global var that's only used once.

Reply:
And actually `empty` is only used once, just replace it with `reflect.DeepEqual(foo, &resource{})`